plugins/mcreator-core/blockly/js/extensions_procedure_deps.js

```diff
@@ -0,0 +1,225 @@
+            const group = Blockly.Events.getGroup();
+            // Makes it so the block change and the unplug event get undone together.
+            Blockly.Events.setGroup(changeEvent.group);
+            const procedure = this.getFieldValue('procedure');
+            for (let i = 0; this.getField('name' + i); i++) {
+                const prevType = this.getInput('arg' + i).connection.getCheck();
+                let depType = null;
+                const depList = javabridge.getDependencies(procedure);
+                for (const dep of depList) {
+                    if (dep.getName() === this.getFieldValue('name' + i)) {
+                        depType = dep.getBlocklyType();
+                        if (depType === '')
+                            depType = [];
+                        break;
+                    }
+                }
+                // Set input checks from dependency type
+                this.getInput('arg' + i).setCheck(depType);
+                const newType = this.getInput('arg' + i).connection.getCheck();
+                // Fire change event if block existed earlier and previous input type was different
+                if (changeEvent.type === Blockly.Events.BLOCK_CHANGE &&
+                    JSON.stringify(prevType) !== JSON.stringify(newType)) {
+                    const inputCheckChange = new Blockly.Events.BlockChange(this, null, 'arg' + i, prevType, newType);
+                    inputCheckChange.run = function (forward) {
+                        const block = this.blockId && this.getEventWorkspace_().getBlockById(this.blockId);
+                        if (block)
+                            block.getInput(this.name).setCheck(forward ? this.newValue : this.oldValue);
+                    };
+                    Blockly.Events.fire(inputCheckChange);
+                }
+            }
+            Blockly.Events.setGroup(group);
+        });
+    });
+
+// Helper function to use in Blockly extensions that validate repeating fields' values meant to be unique
+// The nullValue function is used when mutator needs to set a valid value in the field right after its creation
+function uniqueValueValidator(fieldName) {
+    return function (newValue) {
+        for (let i = 0; this.sourceBlock_.getField(fieldName + i); i++) {
+            if (this.sourceBlock_.getFieldValue(fieldName + i) === newValue && (fieldName + i) !== this.name)
+                return null;
+        }
+        return newValue;
+    };
+}
+
+// Helper function to find first index not taken by repeating fields with a certain name
+function firstFreeIndex(block, fieldName, index, valueProvider) {
+    const values = [];
+    for (let i = 0; block.getField(fieldName + i); i++) {
+        if (index && i === index)
+            continue;
+        values.push('' + block.getFieldValue(fieldName + i));
+    }
+    let retVal = 0;
+    while (true) {
+        if (values.indexOf('' + (valueProvider ? valueProvider(retVal) : retVal)) === -1)

```

Comment:
Checking code, valueProvider always seems to be defined so I would omit this check, as well as at the return call

Reply:
I would say I always try to provide more than is currently needed by the core (same with copy buttons on `JStateLabel`). This one is to allow to shorten code when one doesn't need to add any prefix or suffix to the supplied number.

Reply:
And index parameter nullability is supposed to function like `UniqueNameValidator#setIsPresentOnList(boolean)` in the Java part.

Reply:
If something is not needed at that point yet, unless explicitly asked, it is better to just prepare it in a way it can be refactored easily later.

This function is only used here now so it can be simplified to the point of needs to that file